py4DSTEM/file/datastructure/datacube.py

```diff
@@ -211,5 +230,192 @@ def get_virtual_image_annular_CoMY(self,slice_x,slice_y,R):
+
+    Accepts:
+        electrons:      (PointListArray or h5py.Dataset) array of lists of 
+                        electron strike events. *DO NOT MODIFY THIS AFTER CREATION*
+        detector_shape: (list of ints) size Q_Nx and Q_Ny of detector
+        index_keys:     (list) if the data arrays in electrons are structured, specify
+                        the keys that correspond to the electron data. If electrons
+                        is unstructured, pass [None] (as a list!)
+        use_dask:       (bool) by default, the CountedDataCube.data object DOES NOT
+                        SUPPORT slicing along the realspace axes (i.e. you can ONLY
+                        pass single scan positions). By setting use_dask = True, 
+                        a Dask array will be created that enables all slicing modes
+                        supported by Dask. This can add substantial overhead.
+    """
+
+    def __init__(self,electrons,detector_shape,index_keys='ind',
+                    use_dask=False, **kwargs):
+        DataObject.__init__(self,**kwargs)
+
+        self.electrons = electrons
+        self.detector_shape = detector_shape
+
+        if use_dask:
+            import dask.array as da
+            sa = Sparse4D(self.electrons,detector_shape,index_keys,**kwargs)
+            self.data = da.from_array(sa,chunks=(1,1,detector_shape[0],detector_shape[1]))
+        else:
+            self.data = Sparse4D(self.electrons,detector_shape,index_keys,**kwargs)
+
+        self.R_Nx = int(self.data.shape[0])
+        self.R_Ny = int(self.data.shape[1])
+        self.Q_Nx = int(self.data.shape[2])
+        self.Q_Ny = int(self.data.shape[3])
+
+        self.R_N = self.R_Nx * self.R_Ny
+
+    def bin_data_diffraction(self, bin_factor):
+        # bin the underlying data (keeping in sparse storage)
+        pass
+
+    def bin_data_real(self, bin_factor):
+        # bin the underlying data (keeping sparse storage)
+        pass
+
+    def densify(self,bin_R=1, bin_Q=1, memmap=False, dtype=np.uint16):
+        """
+        Convert to a fully dense DataCube object, with 
+        optional binning in real and reciprocal space.
+        If memmap=True, the dense DC will be stored on disk
+        in a temporary file (using numpy).
+        """
+        newRx = int(np.ceil(self.R_Nx / bin_R))
+        newRy = int(np.ceil(self.R_Ny / bin_R))
+        newQx = int(np.ceil(self.Q_Nx / bin_Q))
+        newQy = int(np.ceil(self.Q_Ny / bin_Q))
+
+        if memmap:
+            #make temp file
+            tf = TemporaryFile()
+            data4D = np.memmap(tf,dtype,'r+',shape=(newRx,newRy,newQx,newQy))
+        else:
+            data4D = np.zeros((newRx,newRy,newQx,newQy),dtype=dtype)
+
+        for (Rx, Ry) in tqdmnd(self.R_Nx, self.R_Ny, desc="Creating dense DC"):
+            rx = Rx // bin_R
+            ry = Ry // bin_R
+
+            data4D[rx,ry,:,:] += bin2D(self.data[Rx,Ry,:,:],bin_Q,dtype=dtype)
```

Comment:
Consider a conditional statement here that skips running bin2D() if bin_q==1

Reply:
👍